### PR TITLE
add blob-capable data source

### PR DESCRIPTION
**Pull request by roberto-bayardo**
**Description**

Adds a data source iterator that is aware of blob batcher transactions that can be used post-Ecotone when the batcher can start using blobs.

**Tests**

Unit tests for confirming blob batcher transactions are appropriately detected and included when valid.

**Additional context**

### CODE REVIEW

File: op-node/rollup/derive/data_source.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+// DataSourceConfig regroups the mandatory rollup.Config fields needed for DataFromEVMTransactions.
+type DataSourceConfig struct {
+	l1Signer          types.Signer
+	batchInboxAddress common.Address
+}
+
+func isValidBatchTx(tx *types.Transaction, l1Signer types.Signer, batcherAddr common.Address) bool {
+	seqDataSubmitter, err := l1Signer.Sender(tx) // optimization: only derive sender if To is correct
```
Comment (tynes) [inline op-node/rollup/derive/data_source.go:72]: This comment is far away from its context now as the `tx.To` check is far away now. Thoughts on moving the `tx.To` check in here?

File: op-node/rollup/derive/blob_data_source.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+		}
+		// handle non-blob batcher transactions by extracting their calldata
+		if tx.Type() != types.BlobTxType {
+			calldata := eth.Data(tx.Data())
+			data = append(data, blobOrCalldata{nil, &calldata})
+			continue
+		}
+		// handle blob batcher transactions by extracting their blob hashes, ignoring any calldata.
+		if len(tx.Data()) > 0 {
+			log.Warn("blob tx has calldata, which will be ignored")
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source.go]: Perhaps also include tx hash in the logline?

File: op-node/rollup/derive/blob_data_source.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+func dataAndHashesFromTxs(txs types.Transactions, config *DataSourceConfig, batcherAddr common.Address) ([]blobOrCalldata, []eth.IndexedBlobHash) {
+	data := []blobOrCalldata{}
+	var hashes []eth.IndexedBlobHash
+	blobIndex := 0 // index of each blob in the block's blob sidecar
+	for _, tx := range txs {
+		// skip any non-batcher transactions
+		if to := tx.To(); to == nil ||
+			*to != config.batchInboxAddress ||
+			!isValidBatchTx(tx, config.l1Signer, batcherAddr) {
+			blobIndex += len(tx.BlobHashes())
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source.go:125]: There are a lot of different cases that need tests that trigger this branch. Perhaps listing off invariants would be useful for ensuring correctness

File: op-node/rollup/derive/data_source.go (hunk 1 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/data_source.go:72]: Ah yes good catch.  Also just noticed I forgot to make the calldata_source.go code use this function too. Both issues fixed.

File: op-node/rollup/derive/blob_data_source.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:125]: moved these checks into isValidTx and documented what it's checking for in the godoc for the func.

File: op-node/rollup/derive/blob_data_source.go (hunk 2 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go]: done

File: op-node/rollup/derive/blob_data_source.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:125]: (and FWIW the unit test I added exercises each of these conditions)

File: op-node/node/node.go [hunk 4]
```diff
@@ -288,6 +293,21 @@ func (n *OpNode) initRuntimeConfig(ctx context.Context, cfg *Config) error {
...
+	httpClient, err := cfg.Beacon.Setup(ctx, n.log)
+	if err != nil {
+		return fmt.Errorf("failed to setup L2 execution-engine RPC client: %w", err)
+	}
+
+	cl := sources.NewL1BeaconClient(httpClient)
+	n.beacon = cl
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (tynes) [reply op-node/node/node.go]: This is a valid find, thank you rabbit. The error message should be updated to refer to the beacon client and not the execution client

File: op-node/node/client.go [hunk 5]
```diff
@@ -164,3 +169,38 @@ func (cfg *PreparedL1Endpoint) Check() error {
...
+
+func (cfg *L1BeaconEndpointConfig) Check() error {
+	if cfg.BeaconAddr == "" {
+		return errors.New("expected beacon address, but got none")
+	}
+	return nil
+}
+
+// PreparedL1BeaconEndpoint enables testing with an in-process pre-setup connection to a beacon node mock
+type PreparedL1BeaconEndpoint struct {
```
Comment (tynes) [inline op-node/node/client.go]: Is this just useful for testing? If so, should it live in `op-e2e`? Or is there prior art in the codebase for having it here?

File: op-node/node/config.go [hunk 6]
```diff
@@ -124,6 +126,11 @@ func (cfg *Config) Check() error {
 	if err := cfg.L2.Check(); err != nil {
 		return fmt.Errorf("l2 endpoint config error: %w", err)
 	}
+	if cfg.Beacon != nil {
```
Comment (tynes) [inline op-node/node/config.go:129]: At some point, this cannot be `nil`. Should we enforce that its not `nil` if ecotone is active to prevent a panic on misconfiguration?

File: op-node/node/node.go [hunk 7]
```diff
@@ -288,6 +293,21 @@ func (n *OpNode) initRuntimeConfig(ctx context.Context, cfg *Config) error {
 	return nil
 }
 
+func (n *OpNode) initL1BeaconAPI(ctx context.Context, cfg *Config) error {
+	if cfg.Beacon == nil {
```
Comment (tynes) [inline op-node/node/node.go:297]: This could create a silent error and possible panic downstream on misconfiguration?

File: op-node/rollup/derive/data_source.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+	if to == nil || *to != batchInboxAddr {
+		return false
+	}
+	seqDataSubmitter, err := l1Signer.Sender(tx) // optimization: only derive sender if To is correct
+	if err != nil {
+		log.Warn("tx in inbox with invalid signature", "hash", tx.Hash(), "err", err)
+		return false
+	}
+	// some random L1 user might have sent a transaction to our batch inbox, ignore them
+	if seqDataSubmitter != batcherAddr {
```
Comment (tynes) [inline op-node/rollup/derive/data_source.go:78]: Would it be useful to also log the sender here? I suppose 1 step of indirection to look up the value on etherscan

File: op-node/rollup/derive/blob_data_source.go [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+// fillBlobPointers goes back through the data array and fills in the pointers to the fetched blob
+// bodies. There should be exactly one placeholder blobOrCalldata element for each blob, otherwise
+// error is returned.
+func fillBlobPointers(data []blobOrCalldata, blobs []*eth.Blob) error {
+	blobIndex := 0
+	for i := range data {
+		if data[i].calldata != nil {
+			continue
+		}
+		if blobIndex >= len(blobs) {
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source.go:160]: Thoughts on defining these errors at the top of the file as variables and then wrapping them with contextual info using `fmt.Errorf`? Then we can test for specific errors in this function

File: op-node/rollup/derive/blob_data_source_test.go [hunk 10]
```diff
@@ -0,0 +1,139 @@
+package derive
+
+import (
+	//"context"
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source_test.go]: Some dead comments in imports

File: op-node/rollup/derive/blob_data_source.go (hunk 9 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:160]: none of these errors should ever be triggered unless someone introduces a logic bug in other parts of the implementation, so not sure it's worth it?  Happy to do it though if you think it's valuable.

File: op-node/rollup/derive/blob_data_source_test.go (hunk 10 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source_test.go]: removed.

File: op-node/node/node.go (hunk 7 above)
Comment (roberto-bayardo) [reply op-node/node/node.go:297]: it would certainly pose a problem if L1 beacon wasn't specified after the Ecotone upgrade. Not sure what the best way to handle this is, as I don't want to break current deploys that aren't even aware of the pending upgrade and won't be specifying the beacon client?

File: op-node/rollup/derive/data_source.go (hunk 8 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/data_source.go:78]: done

File: op-node/node/client.go (hunk 5 above)
Comment (roberto-bayardo) [reply op-node/node/client.go]: this was for the e2e test which will be another PR, so removed from this one.

File: op-node/node/config.go (hunk 6 above)
Comment (roberto-bayardo) [reply op-node/node/config.go:129]: made it error out if ecotone upgrade is active based on current time.

File: op-node/node/node.go (hunk 7 above)
Comment (ajsutton) [reply op-node/node/node.go:297]: Had a similar situation in Teku with the merge where it started requiring an execution engine endpoint - we wound up making the endpoint required if the merge upgrade was scheduled for the current network (even if its not yet active).  That way you can upgrade without changes if you're using a network that doesn't have Ecotone scheduled yet, but when you update to a version that is "Ecotone-ready" it fails to start unless you provide the endpoint to prevent you from thinking you're ready when you're not.

If you only require it when Ecotone is actually active, people will think they're ready and then have it fail when the upgrade actually upgrades.  They're paying more attention when they do an upgrade so that's the time to make it fail. :)

Comment (roberto-bayardo) [reply op-node/node/node.go:297]: added an error log message here to make them aware of the pending upgrade if it's not specified.

Comment (roberto-bayardo) [reply op-node/node/node.go:297]: ok changed based on @ajsutton feedback to error out if the upgrade is scheduled but no endpoint is specified

File: op-node/node/config.go (hunk 6 above)
Comment (roberto-bayardo) [reply op-node/node/config.go:129]: also added a bit more error checking around nil beacon client

File: op-node/node/node.go (hunk 7 above)
Comment (tynes) [reply op-node/node/node.go:297]: Good feedback @ajsutton!

**Comment by tynes** [comment]
This generally looks good to me but needs review from @trianglesphere @axelKingsley @EvanJRichard

File: op-node/rollup/derive/blob_data_source.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+		if len(tx.Data()) > 0 {
+			log.Warn("blob tx has calldata, which will be ignored", "txhash", tx.Hash())
+		}
+		for _, h := range tx.BlobHashes() {
+			idh := eth.IndexedBlobHash{
+				Index: uint64(blobIndex),
+				Hash:  h,
+			}
+			hashes = append(hashes, idh)
+			data = append(data, blobOrCalldata{nil, nil}) // will fill in blob pointers after we download them below
```
Comment (tynes) [inline op-node/rollup/derive/blob_data_source.go:144]: Abstraction feels a bit weird here that its a `nil, nil` used and then filled in outside of the function

**Comment by tynes** [comment]
@clabby These are the derivation pipeline changes for context around the modifications to the fpp

File: op-node/node/node.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (roberto-bayardo) [reply op-node/node/node.go]: fixed

File: op-node/rollup/derive/blob_data_source.go (hunk 11 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:144]: agree it's unusual; conceptually it's just a part of the open() call, but I wanted to be able to unit test it separately, and a fancier abstraction didn't seem to be worth it. Open to ideas here though.

File: op-node/rollup/derive/blob_data_source.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+	if err != nil {
+		ds.log.Error("ignoring blob due to parse failure", "err", err)
+		return ds.Next(ctx)
+	}
+	return data, nil
+}
+
+// open fetches and returns the blob or calldata (as appropriate) from all valid batcher
+// transactions in the referenced block. Returns an empty (non-nil) array if no batcher
+// transactions are found. Returns nil array whenever an error is returned.
```
Comment (sebastianst) [inline op-node/rollup/derive/blob_data_source.go]: Not sure if the function specification should mention this non-nil vs nil slice behavior. It's safer to treat non-nil empty and nil slices semantically the same. I can see that `open` is only called above in `Next`, so we don't rely on this distinction either.

File: op-node/rollup/derive/blob_data_source.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+		// by failing over to a blob archival service.
+		return nil, NewResetError(fmt.Errorf("%w: failed to fetch blobs", err))
+	} else if err != nil {
+		return nil, NewTemporaryError(fmt.Errorf("%w: failed to fetch blobs", err))
+	}
+
+	// go back over the data array and populate the blob pointers
+	if err := fillBlobPointers(data, blobs); err != nil {
+		// this shouldn't happen unless there is a bug in the blobs fetcher
+		return nil, NewResetError(fmt.Errorf("%w: failed to fill blob pointers", err))
```
Comment (sebastianst) [inline op-node/rollup/derive/blob_data_source.go]: Would prefer to suffix rather than prefix the wrapped error (multiple places in this file).

File: op-node/node/client.go [hunk 14]
```diff
@@ -29,6 +29,11 @@ type L1EndpointSetup interface {
 	Check() error
 }
 
+type L1BeaconEndpointSetup interface {
```
Comment (sebastianst) [inline op-node/node/client.go:32]: What's the abstraction for? Is this gonna be important in a different PR?

File: op-node/rollup/derive/blob_data_source.go (hunk 11 above)
Comment (trianglesphere) [reply op-node/rollup/derive/blob_data_source.go:144]: I agree slightly weird, but after reading the flow, it does make some sense to split up finding the blobs from fetching the blobs. I think if we wanted to clean it up, we'd want to make blobOrCalldata be something like the following rust enum or maybe have an internal enum then transform it to an external enum.

```
enum blobOrCalldata {
    Calldata(Vec<u8>),
    UnInitBlob,
    InitBlob(...),
}
```

**Review by trianglesphere — APPROVED** [review]
I'm fine with these changes. We could change the `nil, nil` thing in the middle of `open` if we wanted, but I don't need that change to merge it.

File: op-node/node/node.go (hunk 7 above)
Comment (EvanJRichard) [reply op-node/node/node.go:297]: >ok changed based on @ajsutton feedback to error out if the upgrade is scheduled but no endpoint is specified

It looks to me like the code still doesn't error out (it returns `nil` `error`) and just logs the error and continues - do I misunderstand?

Comment (EvanJRichard) [reply op-node/node/node.go:297]:
```
if cfg.Beacon == nil {
		n.log.Error("No beacon endpoint configured. Configuration is mandatory for the Ecotone upgrade.")
		return nil
	}
```

File: op-node/rollup/derive/blob_data_source.go
Comment (EvanJRichard) [inline op-node/rollup/derive/blob_data_source.go:1]: I really appreciate the heavy commenting in this file.

File: op-node/rollup/derive/l1_retrieval.go [hunk 15]
```diff
@@ -11,7 +12,7 @@ import (
 )
 
 type DataAvailabilitySource interface {
-	OpenData(ctx context.Context, id eth.BlockID, batcherAddr common.Address) DataIter
+	OpenData(ctx context.Context, ref eth.L1BlockRef, batcherAddr common.Address) (DataIter, error)
```
Comment (EvanJRichard) [inline op-node/rollup/derive/l1_retrieval.go:15]: I think this move to using refs over IDs is a good change.

**Review by EvanJRichard — APPROVED** [review]
LGTM too.[ One question in a resolved conversation,](https://github.com/ethereum-optimism/optimism/pull/8778/files#r1445029114) but I consider it non-blocking.

File: op-node/node/node.go (hunk 7 above)
Comment (roberto-bayardo) [reply op-node/node/node.go:297]: thank you, not sure what happened there. fixed.

Comment (roberto-bayardo) [reply op-node/node/node.go:297]: [edit]  This is done in the "check" function in config.go. 
https://github.com/ethereum-optimism/optimism/blob/38254f0bcc7e83f604913cdcc5cfad46699121da/op-node/node/config.go#L134

File: op-node/node/client.go [hunk 16]
```diff
@@ -164,3 +169,20 @@ func (cfg *PreparedL1Endpoint) Check() error {
 
 	return nil
 }
+
+type L1BeaconEndpointConfig struct {
+	BeaconAddr string // Address of L1 User Beacon-API endpoint to use (beacon namespace required)
```
Comment (axelKingsley) [inline op-node/node/client.go:176]: If a specific format is required (beacon namespace required), might be valuable to include an example

File: op-node/node/client.go [hunk 17]
```diff
@@ -164,3 +169,20 @@ func (cfg *PreparedL1Endpoint) Check() error {
...
+func (cfg *L1BeaconEndpointConfig) Setup(ctx context.Context, log log.Logger) (cl client.HTTP, err error) {
+	return client.NewBasicHTTPClient(cfg.BeaconAddr, log), nil
+}
+
+func (cfg *L1BeaconEndpointConfig) Check() error {
+	if cfg.BeaconAddr == "" {
+		return errors.New("expected beacon address, but got none")
+	}
+	return nil
+}
```
Comment (axelKingsley) [inline op-node/node/client.go:190]: Related to previous comment, if namespace is required, can/should this Check function validate that?

File: op-node/node/node.go [hunk 18]
```diff
@@ -288,6 +293,22 @@ func (n *OpNode) initRuntimeConfig(ctx context.Context, cfg *Config) error {
...
+		n.log.Error("No beacon endpoint configured. Configuration is mandatory for the Cotone upgrade")
```
Comment (axelKingsley) [inline op-node/node/node.go]: `Cotone` -> `Ecotone`?

File: op-node/rollup/derive/blob_data_source.go [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+		fetcher:      fetcher,
+		log:          log.New("origin", ref),
+		batcherAddr:  batcherAddr,
+		blobsFetcher: blobsFetcher,
+	}
+}
+
+// Next returns the next piece of batcher data if any remains. It returns ResetError if it cannot
+// find the referenced block or a referenced blob, or TemporaryError for any other failure to fetch
+// a block or blob.
```
Comment (axelKingsley) [inline op-node/rollup/derive/blob_data_source.go]: This function returns the error from `ds.open(ctx)` or returns an `io.EOF`, but I don't see reference to these `ResetError` or `TemporaryError`. Does this comment belong on the internal `ds.open`?

**Review by tynes — APPROVED** [review]

File: op-node/node/client.go (hunk 16 above)
Comment (roberto-bayardo) [reply op-node/node/client.go:176]: this is just saying that engine and eth namespace needs to enabled on the beacon port API. clarified.

File: op-node/node/node.go (hunk 18 above)
Comment (roberto-bayardo) [reply op-node/node/node.go]: yup fixed.

File: op-node/rollup/derive/blob_data_source.go (hunk 19 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go]: I think it's important to document all possible error return values in the godoc of public API functions such as this, so really it belongs on both. I have clarified by explaining when io.EOF is returned from here, and copied the TemporaryError and ResetError documentation to the godoc for the internal open() call.

File: op-node/rollup/derive/blob_data_source.go (hunk 11 above)
Comment (sebastianst) [reply op-node/rollup/derive/blob_data_source.go:144]: I experimented with replacing the union type with an interface [here](https://github.com/ethereum-optimism/optimism/commit/f1b00980064b3b794df092957eb6104ac12bb77e). Not sure if it actually improves things or makes them less readable. It basically removes 1 or 2 `if`s by introducing abstraction.
It feels like it's only half-way there. A proper way would probably be to create a struct around the `[]dataItem` that counts the blob items as they are created and filled.
Feel free to just ignore it ;)

File: op-node/rollup/derive/blob_data_source.go (hunk 9 above)
Comment (sebastianst) [reply op-node/rollup/derive/blob_data_source.go:160]: Agree with @roberto-bayardo here, these are unexpected errors where no special error handling logic would be implemented at the calling site, so not good candidates for dedicated error types.

File: op-node/rollup/derive/blob_data_source.go (hunk 19 above)
Comment (axelKingsley) [reply op-node/rollup/derive/blob_data_source.go]: I appreciate the thoroughness :D. In the event that the error return behavior of the underlying `ds.open` ever changes, we'd need to remember to update this comment at the same time. Not really sure how to make sure that happens, but it seems unlikely to become a problem.

File: op-node/rollup/derive/blob_data_source.go (hunk 12 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go]: removed. I usually like to document where nil values might be returned to avoid panics but for slices I guess the expectation is nil might always be returned instead of the 0 value.

File: op-node/rollup/derive/blob_data_source.go (hunk 13 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go]: fixed.

File: op-node/node/client.go (hunk 14 above)
Comment (roberto-bayardo) [reply op-node/node/client.go:32]: Not sure TBH.  Was just following the pattern of L1 and L2 endpoint setups. Should I remove?

File: op-node/rollup/derive/blob_data_source.go (hunk 12 above)
Comment (sebastianst) [reply op-node/rollup/derive/blob_data_source.go]: Thanks! Yes Go has this thing where it treats nil vs empty slices the same.

File: op-node/rollup/derive/blob_data_source.go (hunk 11 above)
Comment (roberto-bayardo) [reply op-node/rollup/derive/blob_data_source.go:144]: Yeah agree it only feels half-way there.  Let's revisit later?

Comment (sebastianst) [reply op-node/rollup/derive/blob_data_source.go:144]: yep good idea

File: op-node/node/client.go (hunk 14 above)
Comment (sebastianst) [reply op-node/node/client.go:32]: If it keeps consistent with other patterns, happy to leave it in. Up to you.

**Review by sebastianst — APPROVED** [review]
LGTM. Feel free to resolve the last open conversation if ok for you.

File: op-node/node/client.go (hunk 14 above)
Comment (sebastianst) [reply op-node/node/client.go:32]: https://github.com/ethereum-optimism/client-pod/issues/451
